src/raw/mod.rs

```diff
@@ -0,0 +1,1849 @@
+/// assert!(iter.next().is_none());
+/// # Ok::<(), RawError>(())
+/// ```
+///
+/// Individual elements can be accessed using [`docbuf.get(&key)`](Doc::get), or any of
+/// the `get_*` methods, like [`docbuf.get_object_id(&key)`](Doc::get_object_id), and
+/// [`docbuf.get_str(&str)`](Doc::get_str).  Accessing elements is an O(N) operation,
+/// as it requires iterating through the document from the beginning to find the requested
+/// key.
+///
+/// ```
+/// # use bson::raw::{DocBuf, RawError};
+/// let docbuf = DocBuf::new(b"\x13\x00\x00\x00\x02hi\x00\x06\x00\x00\x00y'all\x00\x00".to_vec())?;
+/// assert_eq!(docbuf.get_str("hi")?, Some("y'all"));
+/// # Ok::<(), RawError>(())
+/// ```
+#[derive(Clone, Debug)]
+pub struct DocBuf {
+    data: Box<[u8]>,
+}
+
+impl DocBuf {
+    /// Create a new `DocBuf` from the provided `Vec`.
+    ///
+    /// The data is checked for a declared length equal to the length of the Vec,
+    /// and a trailing NUL byte.  Other validation is deferred to access time.
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// let docbuf: DocBuf = DocBuf::new(b"\x05\0\0\0\0".to_vec())?;
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn new(data: Vec<u8>) -> RawResult<DocBuf> {
+        if data.len() < 5 {
+            return Err(RawError::MalformedValue("document too short".into()));
+        }
+        let length = i32_from_slice(&data[..4]);
+        if data.len() as i32 != length {
+            return Err(RawError::MalformedValue("document length incorrect".into()));
+        }
+        if data[data.len() - 1] != 0 {
+            return Err(RawError::MalformedValue(
+                "document not null-terminated".into(),
+            ));
+        }
+        Ok(unsafe { DocBuf::new_unchecked(data) })
+    }
+
+    /// Create a DocBuf from a [bson::Document].
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// use bson::{doc, oid};
+    /// let document = doc! {
+    ///     "_id": oid::ObjectId::new(),
+    ///     "name": "Herman Melville",
+    ///     "title": "Moby-Dick",
+    /// };
+    /// let docbuf: DocBuf = DocBuf::from_document(&document);
+    /// # Ok::<(), RawError>(())
+    /// ```
+    pub fn from_document(doc: &crate::Document) -> DocBuf {
+        let mut data = Vec::new();
+        doc.to_writer(&mut data).unwrap();
+        unsafe { DocBuf::new_unchecked(data) }
+    }
+
+    /// Create a DocBuf from an owned Vec<u8> without performing any checks on the provided data.
+    ///
+    /// ```
+    /// # use bson::raw::{DocBuf, RawError};
+    /// let docbuf: DocBuf = unsafe {
+    ///     DocBuf::new_unchecked(b"\x05\0\0\0\0".to_vec())
+    /// };
+    /// # Ok::<(), RawError>(())
+    /// ```
+    ///
+    /// # Safety
+    ///
+    /// The provided bytes must have a valid length marker, and be NUL terminated.
+    pub unsafe fn new_unchecked(data: Vec<u8>) -> DocBuf {
+        DocBuf {
+            data: data.into_boxed_slice(),
+        }
+    }
```

Comment:
This function doesn't actually do anything unsafe.  I marked it that way out of a suspicion that it might someday in the future, so it's marked to prevent backwards compatibility semver issues, but everything interacts with the underlying slice in a completely safe manner.

Reply:
Interesting; I was wondering why it was marked as unsafe. I think if it ever does need to become unsafe, we would probably prefer to just add a separate method and deprecate this one rather than marking it as unsafe even though it currently doesn't need to be, so I've removed that annotation in my follow-up commit.